controllers/snow_machineconfig_controller.go

```diff
@@ -0,0 +1,110 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/aws"
+	"github.com/aws/eks-anywhere/pkg/providers/snow"
+)
+
+// SnowMachineConfigReconciler reconciles a SnowMachineConfig object
+type SnowMachineConfigReconciler struct {
+	Client        client.Client
+	Log           logr.Logger
+	ClientBuilder aws.ClientBuilder
+	Validator     *snow.Validator
+}
+
+func NewSnowMachineConfigReconciler(client client.Client, log logr.Logger, clientBuilder aws.ClientBuilder, validator *snow.Validator) *SnowMachineConfigReconciler {
+	return &SnowMachineConfigReconciler{
+		Client:        client,
+		Log:           log,
+		ClientBuilder: clientBuilder,
+		Validator:     validator,
+	}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *SnowMachineConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&anywherev1.SnowMachineConfig{}).
+		Complete(r)
+}
+
+// TODO: add here kubebuilder permissions as needed
+func (r *SnowMachineConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	log := r.Log.WithValues("snowMachineConfig", req.NamespacedName)
+
+	// Fetch the SnowMachineConfig object
+	snowMachineConfig := &anywherev1.SnowMachineConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, snowMachineConfig); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(snowMachineConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		patchOpts := []patch.Option{}
+		if reterr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
+		}
+		if err := patchHelper.Patch(ctx, snowMachineConfig, patchOpts...); err != nil {
+			log.Error(reterr, "Failed to patch snowmachineconfig")
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// There's no need to go any further if the SnowMachineConfig is marked for deletion.
+	if !snowMachineConfig.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, snowMachineConfig, log)
+	}
+
+	result, err := r.reconcile(ctx, snowMachineConfig, log)
+	if err != nil {
+		log.Error(err, "Failed to reconcile SnowMachineConfig")
+	}
+	return result, nil
+}
+
+func (r *SnowMachineConfigReconciler) reconcile(ctx context.Context, snowMachineConfig *anywherev1.SnowMachineConfig, log logr.Logger) (_ ctrl.Result, reterr error) {
+	// TODO: need to figure out how to load creds in controller
+	deviceClientMap, err := r.ClientBuilder.BuildSnowAwsClientMap(ctx)
+	if err != nil {
+		log.Error(err, "failed to build aws client map")
+		return ctrl.Result{}, err
+	}
+	// Setting the aws client map on every reconcile based on the secrets at that point of time
+	validator := r.Validator.WithAwsClients(deviceClientMap)
```

Comment:
Yea I was debating between the two but wanted to see if this implementation would work with the assumption that we don't have to worry about thread safety. I think I prefer creating the validator dynamically so that we don't affect the regular snow flow, but will try it out first and see if the passing in the client is better.